usersession/userd/launcher.go

```diff
@@ -170,6 +185,216 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
+}
+
+// verifyDesktopFileLocation checks the desktop file location and access.
+// 1. we only consider desktop files in dirs.SnapDesktopFilesDir
+// 2. the desktop file itself and all directories above it are root owned without group/other write
+// 3. the Exec line has an expected prefix
+func verifyDesktopFileLocation(desktopFile string) error {
+	if !strings.HasPrefix(desktopFile, dirs.SnapDesktopFilesDir) {
+		// We currently only support launching snap applications from desktop files in
+		// /var/lib/snapd/desktop/applications and these desktop files are written by snapd and
+		// considered safe for userd to process. If other directories are added in the future,
+		// verifyDesktopFileLocation() and parseExecCommand() may need to be updated.
+		return fmt.Errorf("only launching snap applications from /var/lib/snapd/desktop/applications is supported")
+	}
+
+	if filepath.Clean(desktopFile) != desktopFile {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	// When running in a test environment dirs.SnapDesktopFilesDir is under /tmp/ and
+	// cannot pass these checks. Unless there's a way to subvert dirs.SnapDesktopFilesDir
+	// this hack should be harmless. Maybe there's a better way to fake it?
+	if !strings.HasPrefix(dirs.SnapDesktopFilesDir, "/tmp/") {
+
+		last := "/"
+		for _, val := range strings.Split(desktopFile, "/") {
+			checkPath := filepath.Join(last, val) // val is "" for root ('/')
+			fileStat, err := os.Stat(checkPath)
+			if err != nil || !(fileStat.Mode().IsDir() || fileStat.Mode().IsRegular()) || (fileStat.Mode().Perm()&0022) != 0 || fileStat.Sys().(*syscall.Stat_t).Uid != 0 {
+				return fmt.Errorf("cannot verify path %q", checkPath)
+			}
+			last = checkPath
+		}
+	}
+
+	return nil
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry and
+// checks that the BAMF_DESKTOP_FILE_HINT is present and refers to the desktop file.
+func readExecCommandFromDesktopFile(desktopFile string) (string, error) {
+	var launch string
+
+	file, err := os.Open(desktopFile)
+	if err != nil {
+		return launch, err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+
+	in_desktop_section := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if line == "[Desktop Entry]" {
+			in_desktop_section = true
+		} else if strings.HasPrefix(line, "[Desktop Action ") {
+			// maybe later we'll add support here
+			in_desktop_section = false
+		} else if strings.HasPrefix(line, "[") {
+			in_desktop_section = false
+		} else if in_desktop_section && strings.HasPrefix(line, "Exec=") {
+			launch = strings.TrimPrefix(line, "Exec=")
+			break
+		}
+	}
+
+	expectedPrefix := fmt.Sprintf("env BAMF_DESKTOP_FILE_HINT=%s /snap/bin/", desktopFile)
+	if !strings.HasPrefix(launch, expectedPrefix) {
+		return "", fmt.Errorf("Desktop file %q has an unsupported 'Exec' value: %q", desktopFile, launch)
+	}
```

Comment:
This checks was requested by @jdstrand: https://github.com/snapcore/snapd/pull/8699#discussion_r432738935

I'm fine implementing whatever consensus can be reached.

Reply:
This is for defense in depth since there are various potential future outcomes where this might be parsing desktop files from /var/lib/snap, /usr, $HOME, etc. Since we currently want only to support launching other snaps with this, having this check means that a change in other assumptions will mean we are required to think through the security implications of this section.

Note, I did suggest that the expectedPrefix be refactored in some manner so this and wrappers/desktop.go could share it, to reduce the maintenance overhead.

Reply:
While I would still prefer the check for defense in depth and believe there is a way to unit test this, I won't block on its recent removal.